### PR TITLE
feat(accounts-db): flush/clean/shrink/purge logic

### PR DESCRIPTION
**Pull request by 0xNineteen**
### Flushing [x]

flush rooted accounts to disk (requirement: consensus to know rooted slots)

input: given list of slots:
- [x] gather accounts which were written in a slot
- [x] create account file
- [x] push account bytes to disk
- [x] update index to point at disk
- [x] add file to `unclean_account_files` - could either contain zero-lamport or old-state

### Cleaning [x]

removes rooted zero-lamport accounts and old account-states (ie, older than a rooted slot)

eg,
- an account exists in a rooted slot N
- the account transitions at slot N+1 (but isnt rooted)
	- new account state will be in the cache with the index pointing to it
- slot N+1 is rooted
	- accounts in slot N+1 is flushed to disk and index now points to disk
	- account_file is added to `unclean_account_files` queue
- all previous states are removed from index and account file `alive_bytes` is updated for the old state’s file

implementation:
- [x] collect account which have been rooted but not cleaned
- [x] iter over each account
- [x] if state.slot() < rooted_slot (both old-account + zero-lamport case): 
	- (will always be in an account file since states < rooted_slot will also be rooted)
		- look up in account-file and reduce the alive_bytes in the account file
		- if alive_bytes < threshold * file_size => add to `shrink_account_file`
- [x] if zero-lamports
	- also remove zero-lamport account from index + current account_file
	- if alive_bytes < threshold * file_size => add to `shrink_account_file`

### Shrinking [x]

periodically read `shrink_account_file` queue
- [x] resize file to its `alive_bytes`
- [x] iterate over all accounts -> if <pubkey, slot> in index, then its alive, and should be copied

### Purging [x]
input: given a slot, remove all the accounts in the cache associated with the slot

- populate map<slot, account_ref> for cached accounts/slots on addition
- on db.put(accounts) 
    - generate account references 
    - put in index 
    - put in cache 
- on flush 
    - remove slot from map 
- on purge 
    - pop slot from map, remove all references 

---

additional changes include using the cli accounts-db config for the accounts-db struct

### CODE REVIEW

File: src/accountsdb/accounts_file.zig
Comment (0xNineteen) [reply src/accountsdb/accounts_file.zig:137]: i agree in certain cases that syntax is more easier to read - but i disagree in this case

File: src/accountsdb/accounts_file.zig
Comment (0xNineteen) [reply src/accountsdb/accounts_file.zig]: df5df96

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig:915]: 5a934de

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig:344]: 5a934de

File: src/accountsdb/accounts_file.zig
Comment (InKryption) [reply src/accountsdb/accounts_file.zig:137]: Would like to hear your thoughts on this more sometime, I'm curious on the details of your opinion.

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig:885]: 7343cd5

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig:1158]: 7fdbdb8

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig:982]: 7fdbdb8

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig:976]: yeah thats a good approach - i feel like the added complexity for this one case isnt worth it in this case though

Comment (InKryption) [reply src/accountsdb/db.zig:977]: Well, I do feel compelled to ask, is there any reason to not just return a tuple directly from this function? The return type is anonymous, meaning it's not used anywhere else, and it appears its only purpose is to be deconstructed.

Comment (0xNineteen) [reply src/accountsdb/db.zig]: 77533d3

**Review by InKryption — DISMISSED** [review]

File: src/accountsdb/accounts_file.zig
Comment (InKryption) [inline src/accountsdb/accounts_file.zig:80]: I'd like to point out: writeIntLittleMem doesn't return any errors, the return type should be changed, and all unnecessary error union return types it requires should be removed.

File: src/accountsdb/accounts_file.zig
Comment (InKryption) [inline src/accountsdb/accounts_file.zig:78]: Since this type doesn't hold any variable length data, using:
```suggestion
        pub fn writeToBuf(self: *const AccountInfo, buf: *[@sizeOf(AccountInfo)]u8) !comptime_int {
```
would allow for fewer failure states statically.

Making the return type `comptime_int` would just require changing the return statement to `return buf.len;`, before which it should be asserted that `offset == buf.len` (if that isn't true something is very wrong).

**Review by InKryption — CHANGES_REQUESTED** [review]

File: src/accountsdb/accounts_file.zig
Comment (0xNineteen) [reply src/accountsdb/accounts_file.zig:78]: compiler doesnt seem to like that approach - i think it requires AccountInfo to also be comptime known

```
src/core/account.zig:63:35: error: unable to resolve comptime value
        offset += try account_info.writeToBuf(buf[offset..]);
                      ~~~~~~~~~~~~^~~~~~~~~~~
src/core/account.zig:63:35: note: argument to function being called at comptime must be comptime-known
src/accountsdb/accounts_file.zig:78:86: note: expression is evaluated at comptime because the function returns a comptime-only type '@typeInfo(@typeInfo(@TypeOf(accountsdb.accounts_file.AccountInFile.AccountInfo.writeToBuf)).Fn.return_type.?).ErrorUnion.error_set!comptime_int'
        pub fn writeToBuf(self: *const AccountInfo, buf: *[@sizeOf(AccountInfo)]u8) !comptime_int {
                                                                                     ^~~~~~~~~~~~
```

File: src/accountsdb/accounts_file.zig
Comment (0xNineteen) [reply src/accountsdb/accounts_file.zig:80]: good point 

ae91d35

File: src/accountsdb/accounts_file.zig
Comment (InKryption) [reply src/accountsdb/accounts_file.zig]: Ah right, you need to make the function returning `comptime_int` an `inline fn`, so the result is implicitly comptime-known.

Comment (InKryption) [reply src/accountsdb/accounts_file.zig]: In this case I'm not sure it's worth the code bloat just for that, so maybe let's not do that unless we can demonstrate it has worthwhile performance benefits. But it would still be good to make `buf` be a static length buffer.

Comment (0xNineteen) [reply src/accountsdb/accounts_file.zig]: true - tried it out and thought calling it wasnt that great - `fcn(buf[offset...])` is really nice compared to `&buf[offset...offset+54]` - added assert statements that the buffer is big enough

6aac1f1

Comment (InKryption) [reply src/accountsdb/accounts_file.zig]: What about just `buf[offset..][0..@sizeOf(AccountInfo)]`?

Comment (0xNineteen) [reply src/accountsdb/accounts_file.zig]: its the same problem of needing to define the sizeOf(AccountInfo) when calling the method unfortunately so would prefer to stick with simple slice

**Review by InKryption — APPROVED** [review]

**Review by dnut — CHANGES_REQUESTED** [review]
Here's my initial round of feedback. I'll come back and revisit db.zig more thoroughly after you've merged the thread safety changes.

**Comment by 0xNineteen** [comment]
merged #167 and will continue the clean/up here

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig:1057]: yeah that was a bug
201dfe7

File: src/accountsdb/index.zig
Comment (0xNineteen) [reply src/accountsdb/index.zig:632]: yeah thats a great point - maybe something to investigate in the future

File: src/accountsdb/index.zig
Comment (0xNineteen) [reply src/accountsdb/index.zig:746]: yeah thats a great point - ill change it to private

File: src/accountsdb/index.zig
Comment (0xNineteen) [reply src/accountsdb/index.zig:734]: 3ccf627

File: src/accountsdb/index.zig
Comment (0xNineteen) [reply src/accountsdb/index.zig:550]: nice - 3ccf627

File: src/accountsdb/accounts_file.zig
Comment (0xNineteen) [reply src/accountsdb/accounts_file.zig]: yeah good point - this code isnt used a lot of places and was more so remains of the first iteration/testing i was doing - changed it to return number_of_accounts

ea92f91

File: src/accountsdb/index.zig
Comment (0xNineteen) [reply src/accountsdb/index.zig]: yeah i didnt like this logic either - changed it to something much simpler in the newest commit (before reading this comment)

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig:500]: my thoughts were this function should only be run at startup and these errors are un-recoverable so we should just panic 

changed to errors though since i agree with you - b144ae7

**Comment by 0xNineteen** [comment]
note: originally i wanted to improve logs/prom-metrics but for the sake of getting the main logic reviewed and merged, i think its ok to keep logs/metrics bare-bones for now and improve in another pr

File: src/accountsdb/db.zig
Comment (InKryption) [inline src/accountsdb/db.zig:494]: Wouldn't this logic be more simply expressed by `if (references.items.len != 0)`? Is there potentially more logic that should be added in between that would require this sort of state tracking? If so, a comment explaining the context would be helpful to avoid bungling that.

File: src/accountsdb/db.zig
Comment (InKryption) [inline src/accountsdb/db.zig:1748]: Why have this extra branch hidden inside the function? This method doesn't interact with any of the other fields in the account_ref, it'd make more sense to me to just accept the payload type of `AccountLocation.File` as a parameter (it's currently an anonymous declaration, would need to give it a name).

**Review by InKryption — COMMENTED** [review]
I have a few notes

File: src/accountsdb/accounts_file.zig
Comment (0xNineteen) [reply src/accountsdb/accounts_file.zig:113]: 904c87a

File: src/accountsdb/accounts_file.zig
Comment (0xNineteen) [reply src/accountsdb/accounts_file.zig:241]: fdc0390

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig:90]: yeah thats the main reason

in general, ive found coding with future changes leads to worse code - and usually the future changes never happen so you are just left with dead code

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig:481]: 8ed4ad4

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig:494]: 72a9e20

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig:1748]: changed to take location as param - 94aabe7

rn we dont use the method for cache but its _very likely_ we will use it in the future so i think its ok to leave it like this for now

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:90]: I generally would prefer to always at least use these kind of errdefers as expressions of intent wrt the acquired resource, but I won't fret over a couple missing noop errdefers.

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig:1748]: I think that's fine then if it is eventually intended, but in that case I think it would be better to mark this as a panic, to indicate that it is unimplemented, and not just a normal failure state that is reachable assuming normal and correct use of the API (the error set is part of the API after all).

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig:90]: it is an interesting idea - ill think more about it too

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig]: thats a good point - 87d2d9b

File: src/accountsdb/db.zig
Comment (InKryption) [reply src/accountsdb/db.zig]: To be clear, my intended meaning was that it's safer to use unqualified error values as switch prongs, because PTR won't be performed - ie, either use compare using `==` against a qualified error value, or switch on unqualified error prongs.

That said, it's fine, won't block on this small detail.

Comment (0xNineteen) [reply src/accountsdb/db.zig]: do you have an example of both scenarios mentioned in the 'ie'? - not 100% on what you mean

Comment (InKryption) [reply src/accountsdb/db.zig]:
```zig
const A = error { Foo };
fn a() A!void { return error.Foo; } // inference is fine, it's type-checked by the return type
// fn b() !void { return A.F00; } // caught by the compiler, there isn't a F00 in A
// fn c() !void {  }
fn d() !void { return error.F00; } // potentially silent killer if you were expecting A.Foo

test {
    if (A.Foo == error.F00) {} // this compiles, the resolved type used for comparison is `error{ Foo, F00 }`
    switch (A.Foo) {
        // error.F00 => {}, // this is a compile error, the type of the prong value must coerce to the operand; error.F00 is not in A, so it will not coerce
        error.Foo => {}, // this is fine unqualified, because it you were to make a typo, you'd get a compile error for the above reason
    }
}
```
Also note: `A.Foo == error.Foo`, because conceptually, errors are all part of one big enum, and error sets are just subsets of the overarching superset. So whether you qualify it only influences the type checking boundary, not the value itself.

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig:1724]: c221588

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig:1681]: 788a7e9

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig]: ah yeah youre right

b09d776

File: src/accountsdb/db.zig
Comment (dnut) [reply src/accountsdb/db.zig:1695]: >  internal state is a pointer which gets invalidated on function returns

Why would it get invalidated on function returns? Isn't it just a pointer to a field in the RwMux?

Are you talking about this field?

https://github.com/Syndica/sig/blob/b09d776a71b1314f36da08b5e81ef1327b3dc7aa/src/sync/mux.zig#L36

initialized here:

https://github.com/Syndica/sig/blob/b09d776a71b1314f36da08b5e81ef1327b3dc7aa/src/sync/mux.zig#L102

This a pointer to the private field in the RwMux struct. As long as the RwMux struct is still alive, the pointer should still be valid. In this case, the RwMux is a field in AccountsDB:

https://github.com/Syndica/sig/blob/b09d776a71b1314f36da08b5e81ef1327b3dc7aa/src/accountsdb/db.zig#L87

My understanding is that the pointer should be valid as long as the AccountsDB struct is still alive.

File: src/accountsdb/db.zig
Comment (dnut) [inline src/accountsdb/db.zig:1102]: why not use the logger?

File: src/accountsdb/db.zig
Comment (dnut) [inline src/accountsdb/db.zig:1157]: Is this necessary? I thought that as soon as you write it to the mmapped memory, that writes it to the virtual filesystem that the os exposes to all processes. In other words, any read of this file will see the data that you've written, regardless of it actually being on the disk. I thought manual syncing is only necessary if you expect the drive to suddenly get disconnected, otherwise for all intents and purposes it behaves as if the data is on the disk. I may have a misconception here but I could have sworn that's how it works.

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig:1102]: 0473e56

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig:1157]: yeah youre right - the original rust implementation used sync which is why it was there but i saw a recent pr  (missing the link unfortunately) saying the same as what you mentioned and removing it

7afd16a

File: src/accountsdb/db.zig
Comment (0xNineteen) [reply src/accountsdb/db.zig:1695]: huh yeah youre right it works fine - i forget where i previously tried this and it failed 

d9d9da8

**Review by dnut — APPROVED** [review]

File: src/sync/mux.zig
Comment (0xNineteen) [reply src/sync/mux.zig:265]: i prefer the methods i used in this pr - its less code and more in line with other code you see (one method to lock + get and one method to unlock) - you also lose zls types when you do `item.get()`

i kept both approaches here because gossip uses the .get way and didnt want to touch it in this pr - but i think we should transition gossip to use this style too and probs change the `get` to a private method (.. im starting to warm up to private methods lol)

File: src/accountsdb/index.zig
Comment (0xNineteen) [reply src/accountsdb/index.zig:350]: great catch 

4ee27f8

File: src/accountsdb/index.zig
Comment (0xNineteen) [reply src/accountsdb/index.zig:333]: 4ee27f8

Comment (0xNineteen) [reply src/accountsdb/index.zig]: 4ee27f8

**Review by dnut — APPROVED** [review]

**Pull request merged**
